File: du-backend/src/main/java/com/dwin/du/entity/person/PersonController.java
package com.dwin.du.entity.person;

import com.dwin.du.entity.person.Request.AddEditPersonRequest;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import lombok.RequiredArgsConstructor;

@RestController
@RequiredArgsConstructor
@RequestMapping("/rm/person")
@Tag(name = "Person Management", description = "APIs for managing persons")
public class PersonController {

  private final PersonService personService;

    @PostMapping("/add")
    @Operation(summary = "Add a new person", description = "Adds a new person to the system.")
    public ResponseEntity<?> addPerson(@RequestBody AddEditPersonRequest request){
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !authentication.isAuthenticated())
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();

        String currentUsername = authentication.getName();
        return personService.addPerson(request, currentUsername);
    }

    @PutMapping("/{personId}/edit")
    @Operation(summary = "Edit a person", description = "Edits an existing person by personId.")
    public ResponseEntity<?> editPerson(@PathVariable int personId, @RequestBody AddEditPersonRequest request) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.editPerson(personId, request, currentUsername);
    }

    @DeleteMapping("/{personId}/remove")
    @Operation(summary = "Remove a person", description = "Removes a person by personId.")
    public ResponseEntity<?> removePerson(@PathVariable int personId) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.removePerson(personId, currentUsername);
    }

    @PutMapping("/{personId}/set-receipts-counts")
    @Operation(summary = "Set receipts counts", description = "Sets the receipts counts for a person by personId.")
    public ResponseEntity<?> setReceiptsCounts(@PathVariable int personId, @RequestParam int receiptsCount) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.setReceiptsCounts(personId, receiptsCount, currentUsername);
    }

    @PutMapping("/{personId}/set-total-amount")
    @Operation(summary = "Set total amount", description = "Sets the total amount for a person by personId.")
    public ResponseEntity<?> setTotaleAmount(@PathVariable int personId, @RequestParam Double totalAmount) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.setTotalAmount(personId, totalAmount, currentUsername);
    }

    @GetMapping("/{personId}")
    @Operation(summary = "Get person by ID", description = "Retrieves a person by their ID.")
    public ResponseEntity<?> getPerson(@PathVariable int personId) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.getPersonById(personId, currentUsername);
    }

    @GetMapping("")
    @Operation(summary = "Get all persons", description = "Retrieves all persons.")
    public ResponseEntity<?> getPersons(){
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.getPersons(currentUsername);
    }


    @GetMapping("/user-person")
    @Operation(summary = "Get person user", description = "Retrieves a person user.")
    public ResponseEntity<?> getUserPerson() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.getUserPerson(currentUsername);
    }

    @GetMapping("/{receiptId}/from-receipt")
    @Operation(summary = "Get all persons from receipt", description = "Retrieves all persons associated with a specific receipt.")
    public ResponseEntity<?> getPersonsReceipt(@PathVariable int receiptId){
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.getPersonsFromReceipt(currentUsername, receiptId);
    }

    @GetMapping("/{productId}/from-product")
    @Operation(summary = "Get all persons from product", description = "Retrieves all persons associated with a specific product.")
    public ResponseEntity<?> getPersonsProduct(@PathVariable int productId) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String currentUsername = authentication.getName();
        return personService.getPersonsFromProduct(currentUsername, productId);
    }
}

File: du-front/DivvyUp.Web/Controllers/PersonController.cs
using DivvyUp.Web.Interface;
using DivvyUp.Web.RequestDto;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace DivvyUp.Web.Controllers
{
    [Route("rm/person/")]
    [ApiController]
    public class PersonController : ControllerBase
    {
        private readonly IPersonService _personService;

        public PersonController(IPersonService personService)
        {
            _personService = personService;
        }


        [HttpPost]
        [Route("add")]
        [Authorize]
        public async Task<IActionResult> Add([FromBody] AddEditPersonRequest request)
        {
            return await _personService.Add(request, User);
        }

        [HttpPut]
        [Route("edit/{personId}")]
        [Authorize]
        public async Task<IActionResult> Edit([FromBody] AddEditPersonRequest request, [FromRoute] int personId)
        {
            return await _personService.Edit(request, personId, User);

        }

        [HttpDelete]
        [Route("remove/{personId}")]
        [Authorize]
        public async Task<IActionResult> Remove([FromRoute] int personId)
        {
            return await _personService.Remove(personId, User);
        }

        [HttpGet]
        [Route("get/{personId}")]
        [Authorize]
        public async Task<IActionResult> GetPerson([FromRoute] int personId)
        {
            return await _personService.GetPerson(personId, User);
        }

        [HttpGet]
        [Route("get/persons")]
        [Authorize]
        public async Task<IActionResult> GetPersons()
        {
            return await _personService.GetPersons(User);
        }

        [HttpGet]
        [Route("get/user-person")]
        [Authorize]
        public async Task<IActionResult> GetUserPerson()
        {
            return await _personService.GetUserPerson(User);
        }

        [HttpPost]
        [Route("get/{receiptId}/from-receipt")]
        [Authorize]
        public async Task<IActionResult> GetPersonFromReceipt([FromRoute] int receiptId)
        {
            return await _personService.GetPersonFromReceipt(receiptId, User);
        }

        [HttpPost]
        [Route("get/{productId}/from-product")]
        [Authorize]
        public async Task<IActionResult> GetPersonFromProduct([FromRoute] int productId)
        {
            return await _personService.GetPersonFromProduct(productId, User);
        }
    }
}
